Tools/CISupport/ews-build/steps.py

```diff
@@ -7178,25 +7232,40 @@ def run(self):
         if num_unexpected_results:
-            logTextJson = self.log_observer_json.getStdout()
-            yield self._addToLog('stdio', f'Checking results database for unexpected results...\n')
-            successful_filter = yield self.filter_results_using_results_db(logTextJson)
-            unexpected_results_after_filter = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
-            if successful_filter and self.was_filtered and unexpected_results_after_filter:  # If there are unexpected results
-                yield self._addToLog('stdio', f'\nSuccessfully filtered results! Updating unexpected_results.json on disk.\n')
-                self.write_unexpected_results_file_to_master()
-            if not successful_filter:
-                yield self._addToLog('stdio', f'\nFailed to consult results database. Falling back to tip-of-tree...\n')
-                # If results db failed, rebuild without changes to verify causation
-                self.build.addStepsAfterCurrentStep([ValidateChange(verifyBugClosed=False, addURLs=False), RevertAppliedChanges(exclude=['new*', 'scan-build-output*']), ScanBuildWithoutChange()])
-                self.result_message = self.createResultMessage()
-                return defer.returnValue(rc)
+            user_removed_tests = self.getProperty('user_removed_tests', [])
+            user_added_tests = self.getProperty('user_added_tests', [])
+
+            results_json = yield self.decode_results_data()
+            if not results_json:
+                return defer.returnValue(False)
+
+            unexpected_passes = self.get_unexpected_tests(results_json, 'passes') or ['Empty']
+            unexpected_failures = self.get_unexpected_tests(results_json, 'failures') or ['Empty']
+
+            if user_added_tests == unexpected_passes and user_removed_tests == unexpected_failures:
+                yield self._addToLog('stdio', 'Skipping second build since all unexpected results come from user modified expectations.\n')
+                unexpected_results_after_filter = True
+            else:
+                self.setProperty('test_passes', unexpected_passes)
+                self.setProperty('test_failures', unexpected_failures)
+                yield self._addToLog('stdio', f'Checking results database for unexpected results...\n')
+                successful_filter = yield self.filter_results_using_results_db(results_json)
+                unexpected_results_after_filter = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
+                if successful_filter and self.was_filtered and unexpected_results_after_filter:
+                    yield self._addToLog('stdio', f'\nSuccessfully filtered results! Updating unexpected_results.json on disk.\n')
+                    self.write_unexpected_results_file_to_master()
```

Comment:
Is `write_unexpected_results_file_to_master` really creating file on the master, or creating locally on the bot. Please verify.

Reply:
Yep it creates a file on the master. This function is also pre-existing code that got tested in a previous PR :)
```
bash-5.1$ ls public_html/results/macOS-Safer-CPP-Checks-EWS/cf77edef-15730/scan-build-output/
StaticAnalyzerUnexpectedRegressions  results.html  unexpected-results.html  unexpected_results.json
```